File: 2016/d11.cpp
#include <iostream>
#include <algorithm>
#include <vector>
#include <regex>

#include "../libcpp/io.cpp"
#include "../libcpp/slice.cpp"
#include "../libcpp/parse.cpp"
#include "../libcpp/grids.cpp"
#include "../libcpp/convert.cpp"

using vector_str = std::vector<std::string>;
using vector_int = std::vector<int>;
using vector_char = std::vector<char>;
using vector_float = std::vector<float>;

using namespace std;

#include <map>

int main()
{
    vector_str file = import_file("input/d11.txt");
    vector<vector<string>> all_words;
    vector<vector<int>> floors;
    vector<string> Elements;
    for (string &lines : file) {
        replace_char(lines, '-', ' ');
        replace_char(lines, ',', ' ');
        vector_str words = line_to_words(lines);
        all_words.push_back(words);
    }

    for (vector<string> line : all_words) {
        vector_int floor;
        floors.push_back(floor);
        for (int i=0; i<line.size(); i++) {
            //cout << i << ' ' << line[i] << endl;
            if ((line[i] == "compatible") || (line[i] == "generator") || (line[i] == "generator.")) {
                if (find(Elements.begin(), Elements.end(), line[i-1]) == Elements.end())
                    Elements.push_back(line[i-1]);
                vector<string>::iterator it = find(Elements.begin(), Elements.end(), line[i-1]);
                int index = it - Elements.begin();
                floors.back().push_back((1 + index) * (line[i] == "generator" || line[i] == "generator." ? 1 : -1));
            }
        }
    }
    show_matrix(floors);
    
    //regex word_regex("(\w+)(?:-compatible)? (microchip|generator)");
    
    
  

    return 0;
}

File: 2016/d11.py
import re
import sys
from copy import deepcopy
from itertools import permutations, combinations, chain
sys.path.append("../")
import libpy.polishReverse as pr

elements = []
elevator = 0
floors = []

file = open("input/d11.txt", 'r').read()
pattern = r"(\w+)(?:-compatible)? (microchip|generator)"
for line in file.split('\n'):
    floor = re.findall(pattern, line)   
    floors.append(set())
    for f in floor:
        if (f[0] not in elements):
            elements.append(f[0])
        floors[-1].add(((1 + elements.index(f[0])) * (1 if f[1] == "generator" else -1)))
num_elements = len(elements)

TOP = len(floors)-1

def getState(floors, elevator):
    positions = [[None] * 2 for _ in range(num_elements)]
    for f, floor in enumerate(floors):
        for el in floor:
            positions[abs(el) - 1][0 if el > 0 else -1] = f 
    return (elevator,) + tuple(tuple(x) for x in sorted(positions))


def isWin(current_floors):
    for f, floor in enumerate(current_floors):
        if f < TOP and floor:
            return False
    return True


def isValid(floor):
    generatorExist = any(x > 0 for x in floor)
    microchipUnplu = any(x < 0 and -x not in floor for x in floor)
    return not (generatorExist and microchipUnplu)


def moves(floors, elevator):
    pairs = tuple(combinations(floors[elevator], 2))
    single = tuple(({x} for x in floors[elevator]))
    for direction in [1, -1]:
        new_elevator = elevator + direction
        if 0 <= new_elevator <= TOP:
            payload = chain(pairs, single) if direction > 0 else chain(single, pairs)
            for pay in payload:
                new_floors = deepcopy(floors)
                new_floors[elevator] -= set(pay)
                if not isValid(new_floors[elevator]):
                    continue
                new_floors[new_elevator] |= set(pay)
                if not isValid(new_floors[new_elevator]):
                    continue
                yield new_floors, new_elevator



def BFS():
    seen = set()
    queue = [(elevator, floors, getState(floors, elevator), 0)]
    while queue:
        current_elevator, current_floors, current_state, current_step = queue.pop(0)
        #print(current_floors)
        if (isWin(current_floors)): return current_step
        if (current_state in seen): continue
        
        seen.add(current_state)
        for new_floors, new_elevator in moves(current_floors, current_elevator):
            new_state = getState(new_floors, new_elevator)
            if (new_state not in seen):
                queue.append((new_elevator, new_floors, new_state, current_step + 1))

ans1 = BFS()
print(ans1)

for el in range(2):
    num_elements += 1
    for j in (1, -1):
        floors[0].add(j * (num_elements))

ans2 = BFS()
print(ans2)